### ISSUE TITLE
External API [csp27]

### ISSUE THREAD

**Issue by crazedsanity** [crazedsanity/cs-project#54]
There should be an external API so other programs can automatically file bug reports & such by sending XML to a certain address: it should be able to return an id, or various other information.

The first use of this could very easily be for other installations of cs-project to automatically send bug reports (obviously, the user should be able to configure their system to NOT send these reports).  

AUTOMATED ISSUE CREATION:::
The system should be able to test for existing duplicate bugs; if it is a duplicate, maybe create a note on that issue.
### Build External Interface

Interface should use XML to communicate. I would say the best URL for this API would be "/api".  The submitter could then specific something like "/api/{section}" where
"{section}" indicates a specific target (like one for submission, one for reading project
data, etc).  
## Imported Comments

 [19] Comment (Dan@2007-09-28 18:11:49)
For automated issue creation, think about this (hopefully these are conservative estimates):
Worldwide, there are 200 installations of cs-project.
- 50 are 1.0.0
- 100 are 1.5.2
- 30 are 1.8.0
- 18 are 2.0.0-BETA1
- 1 is 1.0.0-BETA3

There's several bugs in 1.5.1.  After fixing it, several people (for whatever reason) still have that version, and their systems automatically report the bug; even though I've closed it, those reports cause the issue to be re-opened.

Obviously, there would be more than just one bug with all those different versions out there, so a fixed bug could easily get re-opened.

For installations that send an automated bug report, some basic logic should apply:
- open bugs are free to be commented upon.
- each system should only report the same bug once
- reports against closed bugs should be handled differently:
  -- re-opening: submitted issue must have a unique property (i.e. if it's a database error that occurs on v1.0.0-BETA5, then "1.0.0-BETA5" can't be in the listed "versions affected" property).
  
  -- Automated bug reports that don't warrant re-opening will be sent back the last solution given for the issue (since multiple solutions can be made).

[20] Comment (Dan@2007-09-28 18:14:55)
It is _IMPERATIVE_ that each installation have the automated bug reporting as a configuration option: otherwise, _MY_ installation might try to automatically file a bug report with itself.

[21] Comment (Dan@2007-09-28 19:04:07)
It is also very important to be able to secure a user's submission, which may contain sensitive information, against anonymous (and even some registered) users from viewing.  See issue #25.

[68] Comment (Dan@2007-10-04 11:48:13)
Reference [CS-Project / INITIAL RELEASE / Solution email contains bad/unparsed data] for some important information about this.  Consider also this scenario:::

30 hosts encounter issue #595 (just an example), which always contains sensitive info.
- Host 1 submits it, gets notified.
- Hosts 2-30 submit remarks, each time generating an email with the previous remarks (thereby the 30th user gets all sensitive information from the previous 29 remarks).
- A solution is posted, sending all users a list of all the remarks, thereby giving each user receives every other user's sensitive information.

Basically, this means that notes have to be able to be marked as "private", with special rules.  Certain users (not just authenticated ones), like "admins", should be able to view all of these private items, as they would otherwise not be able to attempt any sort of solution.  The user that submitted the private message should be able to view it, but not any private messages created by other users/contacts... make sure this logic is VERY SOLID.
